feat: add experimental support for Ty language server

OpeOginni opened the pull request:
- Introduced OPENCODE_EXPERIMENTAL_LSP_TY flag to conditionally enable Ty language server functionality.
- Implemented Ty language server with support for Python file extensions and virtual environment detection.

Proposed fix for #5345

OpeOginni committed:
feat: add experimental support for Ty language server

- Introduced OPENCODE_EXPERIMENTAL_LSP_TY flag to conditionally enable Ty language server functionality.
- Implemented Ty language server with support for Python file extensions and virtual environment detection.

rekram1-node commented on packages/opencode/src/lsp/index.ts:
```diff
@@ -204,6 +205,11 @@ export namespace LSP {
+      if (extension === ".py" || extension === ".pyi") {
+        if (Flag.OPENCODE_EXPERIMENTAL_LSP_TY && !server.experimental) continue // If experimental flag is enabled and server is not experimental, skip
+        if (!Flag.OPENCODE_EXPERIMENTAL_LSP_TY && server.experimental) continue // If experimental flag is disabled and server is experimental, skip
+      }
```
we may need a cleaner pattern for saying:

"If lsp for language X is enabled, disable the other one"

OpeOginni replied:
True, should this work for all other lsp with similar extensions or just the Python one for now?

OpeOginni replied:
Made some updates, put it in a helper function incase it becomes larger in the future

OpeOginni committed:
feat: implement filtering of experimental LSP servers based on OPENCODE_EXPERIMENTAL_LSP_TY flag cleaner pattern

- Added a function to conditionally enable or disable LSP servers (pyright and ty) based on the experimental flag.

OpeOginni committed:
Merge branch 'dev' into feat/experimental-ty-lsp

OpeOginni committed:
feat: Remove log for when TY is diabled as this should be the default

OpeOginni committed:
Merge branch 'dev' into feat/experimental-ty-lsp

rekram1-node commented:
/review

rekram1-node commented:
the only comment left that I recommend is killing that "experimental" key, ignore all the other agent feedback

OpeOginni commented on packages/opencode/src/lsp/server.ts:
```diff
@@ -361,6 +362,59 @@ export namespace LSPServer {
+    },
+    experimental: true,
+  }
+
```
Makes sense bro-chacho

OpeOginni committed:
refactor: remove experimental flag from LSPServer configuration

OpeOginni committed:
Merge remote-tracking branch 'refs/remotes/origin/feat/experimental-ty-lsp' into feat/experimental-ty-lsp

rekram1-node commented on packages/opencode/src/lsp/server.ts:
```diff
@@ -361,6 +361,58 @@ export namespace LSPServer {
+    id: "ty",
+    extensions: [".py", ".pyi"],
+    root: NearestRoot(["pyproject.toml", "ty.toml", "setup.py", "setup.cfg", "requirements.txt", "Pipfile", "pyrightconfig.json"]),
+    async spawn(root) {
```
isn't this spawning even when the var isn't set?

OpeOginni replied:
Ahh I thought the way the logic goes, the servers only get spawned at after being loaded to the `Instance.state`? Before then the server is removed when the var isn't set. 

But then again i could just add a check there just to be 1000% guaranteed.

OpeOginni replied:
Just like how in `index.ts`

```ts
      if (cfg.lsp === false) {
        log.info("all LSPs are disabled")
        return {
          broken: new Set<string>(),
          servers,
          clients,
          spawning: new Map<string, Promise<LSPClient.Info | undefined>>(),
        }
      }
 ```
 
it can return just empty servers and spawning methods

OpeOginni committed:
feat: add experimental flag for LSP server initialization
